### PR TITLE
Civi::pipe - Add support for background worker-processes with JSON pipes

### PR DESCRIPTION
**Pull request by totten**
Overview
----------------------------------------

Add an entry-point for external scripts to send multiple API calls to a CiviCRM worker-process - while only bootstrapping one time.

Comparison
--------------------------------------
The closest point-of-comparison might be the REST-y API:

* _Similar_: An external script can use either RESTy-style or pipe-style to start Civi, authenticate, send API requests, and receive API responses.
* _Different_: REST-style is optimized for remote callers and uses HTTP. Pipe-style is optimized for local callers and uses STDIN/STDOUT.
* _Different_: In REST-style, you bootstrap Civi+CMS once *for every HTTP request*. In pipe-style, you bootstrap Civi+CMS *once per session* - and the session may have many requests/responses.

Cross-reference
----------------------------------------

* Draft documentation is available at:
    * Dev Docs MR: https://lab.civicrm.org/documentation/docs/dev/-/merge_requests/980
    * Dev Docs Preview: https://think.hm/docs/dev-pipe/framework/pipe/
* The `login` functionality depends on `authx_login()` from ~~#22239~~ #22292. 
* This is one of several steps toward https://lab.civicrm.org/dev/core/-/issues/1304. More discussion about this in "Comments".

Pseudocode
----------------------------------------

Loosely speaking, the `Civi::pipe()` mechanism works like this pseudocode:

```php
class Civi {
  public static function pipe(): void {
    fwrite(STDOUT, json_encode($header));
    while ($line = readline(STDIN)) {
      $request = json_decode($line);
      $result = civicrm_api(...$request);
      fwrite(STDOUT, json_encode($result));
    }
  }
}
```

The actual protocol is a bit more robust - eg it complies with JSON-RPC v2.0, catches errors, supports APIv3/APIv4, and includes some test+control functionality. The protocol is described more thoroughly in [Dev Docs Preview](https://think.hm/docs/dev-pipe/framework/pipe/).

Example
----------------------------------------

To interact with a pipe, you first need to start the `Civi::pipe()` function, eg

```bash
cd /var/www/example.com/web
cv ev 'Civi::pipe("vtl");'
```

This responds with a welcome/header line to describe the Civi environment:

```javascript
{"Civi::pipe":{"v":"5.46.alpha1","t":"trusted","l":["login"]}}
```

The Dev Guide describes the [connection flags and welcome/header](https://think.hm/docs/dev-pipe/framework/pipe/#flags). In short, this header says that:

* The system runs (v)ersion 5.46.alpha1
* The session is (t)rusted for sensitive operations (like disabling API permission-checks).
* The system supports (l)ogins (eg JWT/API key/username-password).

Now that a session is started, we can send and receive method-calls. Method-calls are generally APIv3/APIv4. The calls are formatted with JSON-RPC v2.0. For example:

```javascript
// Send request for `api3("System", "flush")`
{"jsonrpc":"2.0","method":"api3","params":["System","flush",{"check_permissions":0}],"id":null}
// Receive response
{"jsonrpc":"2.0","result":{"is_error":0,"version":3,"count":1,"values":1},"id":null}
```

Additionally, there are some method-calls designed for testing or controlling the session. `echo` is handy way to see if it's working:

```javascript
// Send request for `echo("hello world")`
{"jsonrpc":"2.0","method":"echo","params":["hello world"],"id":null}
// Receive response "hello world"
{"jsonrpc":"2.0","result":["hello world"],"id":null}
```

You can see this in action by manually running `cv ev 'Civi::pipe();'` and then typing/pasting some JSON. 

For E2E testing purposes, the PR includes a basic PHP client which opens a subprocess and encodes/decodes JSON.The above calls would look like this:

```php
$client = new BasicPipeClient('cv ev \'Civi::pipe();\'');
$result = $client->call('api3', ['System', 'flush']);
$result = $client->call('echo', ['hello world']);
```

Comments
----------------------------------------

Why? Let me present the context. In [dev/core#1304](https://lab.civicrm.org/dev/core/-/issues/1304), we want to allow a deployment style wherein a *pool of workers* executes jobs from the queue. How should the pool of function? Generally:

1. A worker should handle several requests - without needing to (re)bootstrap CiviCRM+CMS. This reduces overhead/latency.
2. A worker can be started before a specific queue task is identified. This again reduces overhead/latency.
3. A worker should have a finite lifespan - over time, there is a high risk of memory-leaks being introduced in Civi+CMS code, and the simplest mitigation is to set a quantitative limit on much work each handles. (Ex: "Perform up to 50 tasks over up to 15 minutes.") As a corollary, new workers need to be occasionally started to replace expired workers.
4. A worker may be reused - with boundaries. For example, many background tasks are designed for system-tasks, but some may be designed to execute "on behalf of"  some user. If you have two tasks running "on behalf of"  different users, they should run in separate processes.

This arrangement implies two general components or layers: the *pool manager* and the *worker*. Where should these components live? 

It might be nice to include both manager+worker in `civicrm-core`. Unfortunately, putting the manager  in `civicrm-core` is problematic. It is easiest to develop a manager in a framework like [ReactPHP](https://reactphp.org/) - which is a popular system for asynchronous multiprocessing in PHP. Unfortunately, as a large+popular framework, it's liable to encounter dependency conflicts with other tools in the CMS+Civi ecosystems. Moreover, asynchronous PHP code (like ReactPHP) is stylistically very different from traditional synchronous PHP code (like Drupal/Civi/etc) -- mixing the two styles is likely to invite confusion. So it is best for the manager tier to be a distinct codebase with separate dependencies. (Note: You can see this style of code in `dev/core#1304`s prototypes and in the developmental repo https://github.com/totten/citges - eg [PipePool](https://github.com/totten/citges/blob/master/src/PipePool.php).)

That leaves the worker code. Essentially, the job of a worker is to "go execute an API" or "go execute a queue task". When you get past all the tests + QA + error-handling + permission-checks + standards, it boils down to a loop like this:

```php
while ($line = readline(STDIN)) {
  $result = civicrm_api(...json_decode($line));
  fwrite(STDOUT, json_encode($result));
}
```

And that, basically, is what `Civi::pipe()` does. This works fine as standard PHP and doesn't require ReactPHP (or similar). But it adds in a bunch of tests + QA + error-handling + permission-checks + standards. 

Lastly, there are a few ways you could package-up this worker -- e.g. make it a standalone script; make a script in `extern/`; make a subcommand in `cv`; etc. In point of fact, `cv` does have a somewhat similar command (`cv api:batch`). However, that only works with `cv` -- and many deployers still have occasions to prefer launching CLI commands via `wp`/`drush`. The handy thing about `Civi::pipe()` is that it's easy to adapt to each of these connectors:

```bash
cv ev 'Civi::pipe();'
drush ev 'civicrm_initialize(); Civi::pipe();'
wp eval 'civicrm_initialize(); Civi::pipe();'
```

### CODE REVIEW

File: Civi/Pipe/BasicPipeClient.php [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    if (!$this->process) {
+      throw new \RuntimeException("Client error: Connection was not been opened yet.");
+    }
+
+    $requestLine = json_encode(['jsonrpc' => '2.0', 'method' => $method, 'params' => $params, 'id' => $id]);
+    fwrite($this->pipes[0], $requestLine . "\n");
+    $responseLine = stream_get_line($this->pipes[1], $this->bufferSize, "\n");
+    $decode = json_decode($responseLine, TRUE);
+    if (!isset($decode['jsonrpc']) || $decode['jsonrpc'] !== '2.0') {
+      throw new \RuntimeException("Protocol error: Response lacks JSON-RPC header.");
```
[the remark this answers is not in the document]
Comment (totten) [reply Civi/Pipe/BasicPipeClient.php:125]: @eileenmcnaughton  OK, sent a follow-up https://github.com/civicrm/civicrm-core/pull/22503

**Comment by totten** [comment]
> (@eileenmcnaughton) As noted in conversation I think we should add commands to access civipipe eg `drush cvpipe` or `drush civicrm_pipe`

I pushed up [`cv pipe`](https://github.com/civicrm/cv/pull/110). Should be equally simple to add to drush or wp-cli (it just takes 4x PRs for D7/D8-9/WP/BD).

I do think the shorthand is a little nicer. I was a little concerned about version-hell (ie coordinating the deployment of civicrm-core and cv/drush/wp-cli plugins). Although... on second thought... that won't be a major issue for `drush` or `wp-cli` if we get the shorthand into the same 5.47.alpha.

> Having said that it is pretty painful to use on the command line due to the json input - so it's not likely someone would use this instead of multiple calls to drush cvapi when actually on the command line.

Agree. Aside from some testing/inspection/learning scenarios, I wouldn't expect a human to directly use JSON-RPC. It's not pleasant to write interactively. (Well, I guess it's better than writing ASN.1...)

Tangentially, I found out that you can add GNU readline (*ie arrow-keys and command-history*) via [socat](https://www.redhat.com/sysadmin/getting-started-socat), eg

```bash
socat "EXEC:cv pipe" readline
```

`socat` is wild stuff.

> I gave this a spin and was able to do the specified steps. In terms of security my understanding is this does not expose anything extra except via the command line at this stage.

Yup. This keeps the stakes relatively low. Anyone with command-line/bash/ssh access already has full admin privileges... so giving them a `t`rusted pipe won't really change their access.

Additionally... for bleeding-edge/experimental folk, it is possible to expose pipes as an intranet/TCP service (eg xinetd/sslwrap/socat/etc), but that needs the `u`ntrusted flag; eg

```bash
socat "EXEC:cv pipe uvl" TCP-LISTEN:6002

telnet localhost 6002
# Trying 127.0.0.1...
# Connected to localhost.
# Escape character is '^]'.
# {"Civi::pipe":{"u":"untrusted","v":"5.47.alpha1","l":["nologin"]}}
```

But `u`ntrusted mode is sort of caveat emptor for now.

**Comment by totten** [comment]
>  (@eileenmcnaughton) As noted in conversation I think we should add commands to access civipipe eg `drush cvpipe` or `drush civicrm_pipe`

* https://github.com/civicrm/civicrm-backdrop/pull/154
* https://github.com/civicrm/civicrm-drupal/pull/652
* https://github.com/civicrm/civicrm-drupal-8/pull/70
* https://github.com/civicrm/civicrm-wordpress/pull/270

**Comment by highfalutin** [comment]
This is very exciting to me. Thank you!

**Pull request merged**